dplyr 1.1.0: Joins (#618)
tidyverse/tidyverse.org: content/blog/dplyr-1-1-0-joins/index.Rmd

---
output: hugodown::hugo_document
slug: dplyr-1-1-0-joins
title: "dplyr 1.1.0: Joins"
date: 2023-01-31
author: Davis Vaughan
description: >

>     In dplyr 1.1.0, joins have been greatly reworked, introducing: a new way to

Reviewer: ```suggestion
    In dplyr 1.1.0, joins have been greatly reworked, including a new way to
```

>     specify join columns, various new types of joins, and two new quality

Reviewer: "support for inequality, rolling, and overlap joins" ?

  author: Duy Pham
categories: [package] 
tags: [dplyr]
editor_options: 
  chunk_output_type: console
---

[dplyr 1.1.0](https://dplyr.tidyverse.org/news/index.html#dplyr-110) is out now!
This is post 1 of 4 detailing some of the new features in this release.
In this post, we will discuss various new updates to dplyr's joins.

You can install it from CRAN with:

```{r, eval = FALSE}
install.packages("dplyr")
```

```{r setup, warning=FALSE, message=FALSE}
library(dplyr)
```

## `join_by()`

Consider the following two tables, `transactions` and `companies`.
`transactions` tracks sales across various years for different companies, and `companies` connects the short company id to its actual company name - either Patagonia (a fellow B-Corp!) or RStudio.

```{r}
transactions <- tibble(
  company = c("A", "A", "B", "B"),
  year = c(2019, 2020, 2021, 2023),
  revenue = c(50, 4, 10, 12)
)
transactions

companies <- tibble(
  id = c("A", "B"),
  name = c("Patagonia", "RStudio")
)
companies
```

To join these two tables together, we might use an inner join:

```{r}
transactions |> 
  inner_join(companies, by = c(company = "id"))
```

This works great, but has always felt a little clunky.

> Specifying `c(company = "id")` is a little unnatural for new users, especially if they are used to "equivalence" in R being expressed with `==`.

Reviewer: ```suggestion
Specifying `c(company = "id")` is a little awkward because it uses `=`, not `==`: here we're asserting that we want `company` to equal `id`, not naming a function argument or doing assignment.
```

We've improved on this with a new helper, `join_by()`, which takes expressions in a way that allows you to more naturally express this join:

```{r}
join_by(company == id)
```

This *join specification* can be used as the `by` argument in any of the `*_join()` functions:

```{r}
transactions |> 
  inner_join(companies, by = join_by(company == id))
```

This small quality of life improvement is just one of the many new features that come with `join_by()`.
We'll look at more of these next.

> ## Inequality joins

Reviewer: Maybe this heading should be multiple matches?

Author: Oh duh, great idea thanks

)

companies
```

This table now also tracks name changes that have happened over the course of a company's history.
In 2022, we changed our name from RStudio to Posit, so we've tracked that as an additional row in our dataset.
Note that both RStudio and Posit are given an `id` of `B`, which links back to the `transactions` table.

If we were to join these two tables together, ideally we'd bring over the name that was in effect when the transaction took place.
For example, for the transaction in 2021, the company was still RStudio, so ideally we'd only match up against the RStudio row in `companies`.
If we colored the expected matches, they'd look something like this:

![](img/ideal-join.png)

How can we do this?
We can try the same join from before, but we won't like the results:

```{r}
faulty <- transactions |> 
  inner_join(companies, by = join_by(company == id))

faulty
```

Company `A` matches correctly, but since we only joined on the company id, we get *multiple matches* for each of company `B`'s transactions and end up with more rows than we started with.
This is a problem, as we were expecting a 1:1 match for each row in `transactions`.

> Multiple matches in equality joins like this one are typically unexpected -- in fact, many people don't even know this is possible even though it is technically default SQL behavior -- so we've also added a new warning to alert you when this happens.

Reviewer: ```suggestion
Multiple matches in equality joins like this one are typically unexpected (even though they're baked in to SQL) so we've also added a new warning to alert you when this happens.
```

If multiple matches are expected, explicitly set `multiple = "all"` to silence this warning.
This also serves as a code "sign post" for future readers of your code to let them know that this is a join that is expected to increase the number of rows in the data.
If multiple matches *aren't* expected, you can also set `multiple = "error"` to immediately halt the analysis.
We expect this will be useful as a quality control check for production code where you might rerun analyses with new data on a rolling basis.

> To actually fix this issue, we'll need to expand our join specification to include another condition.

Reviewer: And then add inequality joins heading here?

Since multiple matches are the typical case with inequality joins, we don't get a warning like with the equality join, but we clearly still haven't gotten the join right.
As a reminder, here are where we still have too many matches:

```{r}
transactions |>
  inner_join(companies, join_by(company == id, year >= since)) |>
  filter(company == "B", year == 2023)
```

We need a way to filter down the matches returned from `year >= since` to only the most recent name change.
In other words, we prefer the Posit match over the RStudio match because 2022 is *closer* to the transaction year of 2023 than 2009 is.
We can express this in `join_by()` by using a helper named `closest()`.

```{r}
transactions |>
  inner_join(companies, join_by(company == id, closest(year >= since)))
```

`closest(year >= since)` finds all of the matches in `since` for a particular `year`, and then filters them down to only the closest match to that `year`.
This is known as a *rolling join*, because in this case it *rolls* the most recent name change forward to match up with the transaction.

> Rolling joins were popularized by data.table, and are related to `ASOF` joins supported by some SQL flavors.

Reviewer: ```suggestion
Rolling joins were popularized by data.table, and are related to "as of" joins supported by some SQL flavors.
```
?

Author: I always see it as `ASOF` https://questdb.io/docs/reference/sql/join/#:~:text=ASOF%20JOIN,timestamp%20in%20the%20first%20table.

I always get frustrated when google searching for this because `as of join` doesn't pull anything up, it has to be `asof join` 😆

Reviewer: Then how about `` `ASOF` ("as of") `` because I tried to figure out what the acronym stood for.

> 

Reviewer: Mention overlap joins here? I just that they exist.
